### PR TITLE
Fix #50: Apache Log Format

### LINKED ISSUE #50: Apache input
Technically, it's very similar to Nginx, and the Nginx format can already be used to parse Apache logs.

The only difference is the syntax and we're losing the ability to copy paste the configuration as it is.

Should we duplicate the Nginx format and keep the same spirit for Apache?

Or we just advise developers to use the Nginx format?

### PR DESCRIPTION
**Pull request by benfle**
Basic support for Apache log format.

Do not handle edge cases like custom time formats but is enough to parse logs in the standard combined log format and the Access Watch log format.

See http://httpd.apache.org/docs/current/mod/mod_log_config.html

### CODE REVIEW

File: format/apache.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  const p = compile(format)
+  return (msg) => {
+    return p(msg).reduce((log, v, k) => {
+      const transform = transformers[k.toLowerCase()]
+      if (transform) {
+        return transform(log, v)
+      }
+      const header = k.match(/%\{([^}]+)\}i/)
+      if (header) {
+        const name = header[1].toLowerCase()
```
Comment (znarf) [inline format/apache.js]: Extract the header handling for more readability?

```javascript
const header = key.match(/%\{([^}]+)\}i/)
if (header) {
  return addHeader(log, header[1], value)
}
```

```javascript
function addHeader (log, name, value) {
  name = name.toLowerCase()
  log = log.updateIn(['request', 'captured_headers'], List(), list => list.push(name))
  if (value !== '-') {
    log = log.setIn(['request', 'headers', name], value)
  }
  return log
}
```

File: format/apache.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+function parseTime (value) {
+  value = value.substring(1, value.length - 1) // remove '[' and ']'
+  return strptime(value, '%d/%B/%Y:%H:%M:%S %z').toISOString()
+}
+
+const transformers = {
+  '%h': (log, value) => log.setIn(['request', 'address'], value),
+  '%t': (log, value) => log.setIn(['request', 'time'], parseTime(value)),
+  '%r': addRequest,
```
Comment (znarf) [inline format/apache.js]: Why not use the parseRequest syntax like it was in Nginx?

```javascript
'%r': (log, value) => log.update('request', request => parseRequest(value).merge(request))
```

```javascript
function parseRequest (request) {
  const res = /([^ ]+)\s+([^ ]+)\s+([^ ]+)/.exec(request)
  if (!res) {
    throw new Error(`Could not parse request: '${request}'.`)
  }
  return Map({ method: res[1], url: res[2], protocol: res[3] })
}
```

File: format/apache.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  const parts = format.split(/(%[^ "]+)/)
+
+  const matchString = ([head, ...tail], text) => {
+    if (text.startsWith(head)) {
+      if (tail.length === 0) {
+        return Map()
+      }
+      return matchVariable(tail, text.substring(head.length))
+    } else {
+      throw new Error(`Syntax error. Was expecting the string ${head}. Got: ${text}`)
```
Comment (znarf) [inline format/apache.js]: Move the Exception on top and reduce nesting/indentation?

```javascript
if (!text.startsWith(head)) {
  throw new Error(`Syntax error. Was expecting the variable ${head}. Got ${text}.`)
}
```

File: format/apache.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    }
+  }
+
+  const matchVariable = ([head, ...tail], text) => {
+    const len = text.indexOf(tail[0])
+    if (len !== -1) {
+      const value = text.substring(0, len)
+      return matchString(tail, text.substring(len)).set(head, value)
+    } else {
+      throw new Error(`Syntax error. Was expecting the variable ${head}. Got ${text}.`)
```
Comment (znarf) [inline format/apache.js]: Move the Exception on top and reduce nesting/indentation?

```javascript
if (len === -1) {
  throw new Error(`Syntax error. Was expecting the variable ${head}. Got ${text}.`)
}
```

File: format/apache.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    }
+  }
+
+  return s => {
```
Comment (znarf) [inline format/apache.js]: Could be shortened and make clear what 's' is.

```javascript
return line => matchString(parts, line)
```

File: format/apache.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  '%h': (log, value) => log.setIn(['request', 'address'], value),
+  '%t': (log, value) => log.setIn(['request', 'time'], parseTime(value)),
+  '%r': addRequest,
+  '%>s': (log, value) => log.setIn(['response', 'status'], parseInt(value))
+}
+
+function parser ({format}) {
+  const p = compile(format)
+  return (msg) => {
+    return p(msg).reduce((log, v, k) => {
```
Comment (znarf) [inline format/apache.js]: Extract the reducer for greater readability?

Maybe also, extract baseLog and make the variables clearer.

```javascript
function parser ({format}) {
  const baseLog = fromJS({request: {headers: {}}})
  const lineParser = compileLineParser(format)
  return line => lineParser(line).reduce(reducer, baseLog)
}
```

File: format/apache.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        return log.setIn(['request', 'headers', name], v)
+      }
+      return log
+    }, fromJS({request: {headers: {}}}))
+  }
+}
+
+const formats = {
+  combined: '%h %l %u %t "%r" %>s %b "%{Referer}i" "%{User-agent}i"',
+  accessWatch: '%h %l %u %t "%r" %>s %b "%{Referer}i" "%{User-agent}i" "%{Accept}i" "%{Accept-Charset}i" "%{Accept-Encoding}i" "%{Accept-Language}i" "%{Connection}i" "%{Dnt}i" "%{From}i" "%{Host}i"'
```
Comment (znarf) [inline format/apache.js]: Let's rename accessWatch to accessWatchCombined and support the same format on the Nginx side.

**Comment by benfle** [comment]
@znarf I merged your changes.

**Comment by znarf** [comment]
Great!

**Pull request merged**